pkg/operator/ceph/cluster/mgr/spec_test.go

```diff
@@ -0,0 +1,178 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mgr
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	cephv1beta1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1beta1"
+	rookalpha "github.com/rook/rook/pkg/apis/rook.io/v1alpha2"
+	"github.com/rook/rook/pkg/clusterd"
+	cephconfig "github.com/rook/rook/pkg/daemon/ceph/config"
+	mgrdaemon "github.com/rook/rook/pkg/daemon/ceph/mgr"
+	cephtest "github.com/rook/rook/pkg/operator/ceph/test"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	optest "github.com/rook/rook/pkg/operator/test"
+	testop "github.com/rook/rook/pkg/operator/test"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestPodSpec(t *testing.T) {
+	c := New(
+		&clusterd.Context{Clientset: testop.New(1)},
+		"ns",
+		"rook/rook:myversion",
+		rookalpha.Placement{},
+		false,
+		cephv1beta1.DashboardSpec{},
+		v1.ResourceRequirements{
+			Limits: v1.ResourceList{
+				v1.ResourceCPU: *resource.NewQuantity(100.0, resource.BinarySI),
+			},
+			Requests: v1.ResourceList{
+				v1.ResourceMemory: *resource.NewQuantity(1337.0, resource.BinarySI),
+			},
+		},
+		metav1.OwnerReference{},
+	)
+
+	mgrTestConfig := mgrConfig{
+		DaemonName:   "a",
+		ResourceName: "mgr-a",
+	}
+
+	d := c.makeDeployment(&mgrTestConfig)
+
+	assert.NotNil(t, d)
+	assert.Equal(t, "mgr-a", d.Name)
+	assert.Equal(t, "mgr-a", d.ObjectMeta.Name)
+	assert.Equal(t, 0, len(d.ObjectMeta.Annotations))
+
+	pod := d.Spec.Template
+	assert.Equal(t, appName, pod.ObjectMeta.Labels["app"])
+	assert.Equal(t, c.Namespace, pod.ObjectMeta.Labels["rook_cluster"])
+	assert.Equal(t, 2, len(pod.ObjectMeta.Annotations))
+	assert.Equal(t, "true", pod.ObjectMeta.Annotations["prometheus.io/scrape"])
+	assert.Equal(t, strconv.Itoa(metricsPort), pod.ObjectMeta.Annotations["prometheus.io/port"])
+	assert.Equal(t, v1.RestartPolicyAlways, pod.Spec.RestartPolicy)
+	assert.Nil(t, optest.VolumeExists("rook-data", pod.Spec.Volumes))
+	assert.Nil(t, optest.VolumeExists(cephconfig.DefaultConfigMountName, pod.Spec.Volumes))
+	assert.Nil(t, optest.VolumeExists(k8sutil.ConfigOverrideName, pod.Spec.Volumes))
+
+	assert.Equal(t, 1, len(pod.Spec.InitContainers))
+	assert.Equal(t, 1, len(pod.Spec.Containers))
+
+	configImage := "rook/rook:myversion"
+	configEnvs := 8
+	configContainerDefinition := cephtest.ContainerTestDefinition{
+		Image:   &configImage,
+		Command: []string{}, // no command
+		Args: [][]string{
```

Comment:
Why do the Args need to be of type `[][]string`? If it were just `[]string`, it would seem more readable. Is it to remove any dependence on ordering of the args?

Reply:
I went through a few iterations of the definition for expected args. One of the ones I rejected was making a struct for each individual arg-value(s) set since that was more boilerplate than just using a list-of-lists-of-strings. However, I can revisit this if you think it will improve readability.

Just to make sure there isn't some bit of information I haven't conveyed well, I'll try to explain what I did. The outer layer is a list of args/values. And that list can be in any order. The inner layer is some number of ordered arg-value(s) lists. In the example below, the expected args match the actual args, even though the expected arg-value(s) lists are out of order compared to the actual args. But it still verifies that "a" comes after "--mon" and before "1.2.3.1:6790". And that "12345" must follow "--fsid". etc.
```go
expectedArgs := [][]string {
    {"--mon", "a", "1.2.3.1:6790"},
    {"--fsid", "12345"},
    {"--name", "a"},
}

actualArgs := []string {
    "--name", "a",
    "--fsid", "12345",
    "--mon", "a", "1.2.3.1:6790",
}
```


A list of structs for the same might look like below, but I rejected this due to the fact that the list of strings for args is necessary for the mons in the `monmaptool` init container but it isn't for any other args, so the string list with a single string seemed too much boilerplate for me.
```go
type ExpectedArg struct {
    Flag string
    Args []string
}

expectedArgs := []ExpectedArgs{
    {Flag: "--mon", Args: []string{"a", "1.2.3.1:6790"}},
    {Flag: "--fsid", Args: []string{"12345"}},
    {Flag: "--name", Args: []string{"a"}},
}

...
```

I guess this could be avoided if all the args were in `--flag=value` syntax instead of allowing `[ ... , "--flag", "value" , ... ]`, but I prefer building args with separate flags and values to eliminate the number of `fmt.Sprintf("--flag=%s", value)` statements that are also boilerplate that makes it harder to read the spec.

Reply:
Thanks for the detailed explanation, the benefit is certainly to provide flexible ordering of the args. Other places in the testing we have assumed a constant ordering to keep the test simpler, but now we could increase the usage of this elsewhere as we touch the code.

Reply:
Yeah. My thinking was that the `spec_test.go` files can be updated to use these tests for each of the init refactors. 

And the other backends can use the Args/Volume tests as they like as well.